### PR TITLE
FM2-213:Create update and delete methods for condition

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/FM2-213
WIP  still working on this ticket
cc @ibacher  @corneliouzbett

### CODE REVIEW

**Comment by sherrif10** [comment]
In ConditionFhirResourceProviderResource for r4, some testing methods like JSON_CONDITION_WRONG_UUID which i tried to create ,  breaks some other files, still wondering why condition class still makes a difference than other classes. Again, Are we supposed to create a condition resource web test for r3 ! cc @ibacher  @corneliouzbett ,

**Comment by sherrif10** [comment]
Hello @ibacher  @corneliouzbett , i know there is alot to finish , More suggestions will be highly appreciated thanks

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 1]
```diff
@@ -212,4 +218,14 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
 		assertThat(resultList.size(), greaterThanOrEqualTo(1));
 		assertThat(resultList.iterator().next().fhirType(), equalTo(FhirConstants.CONDITION));
 	}
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:220]: @sherrif10  add unit test for update methods.
Also failure in deletion as well

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 2]
```diff
@@ -212,4 +218,14 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
...
+	
+	@Delete
+	public OperationOutcome deleteObservationResource(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Condition condition = conditionService.delete(id.getIdPart());
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:224]: @sherrif10  we have to mock here not directly call the service method

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 3]
```diff
@@ -213,4 +235,32 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
...
+		
+		Condition wrongCondition = new Condition();
+		wrongCondition.setId(WRONG_CONDITION_UUID);
+		
+		when(conditionService.update(eq(WRONG_CONDITION_UUID), any(Condition.class)))
+		        .thenThrow(MethodNotAllowedException.class);
+		
+		resourceProvider.updateCondition(new IdType().setValue(WRONG_CONDITION_UUID), wrongCondition);
+		verify(conditionService.update(eq(WRONG_CONDITION_UUID), any(Condition.class)));
+
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:250]: I think you have to have the same set of unit tests here as well @sherrif10  as r3

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java (hunk 1 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:220]: ok thanks @Akayeshmantha , will add them

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java (hunk 3 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:250]: sorry for late reply however let me bump into this tomorow thanks

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java (hunk 1 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:220]: Finished that

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java (hunk 2 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:261]: Actually @Akayeshmantha  i tried adding the mock but in resolving stubbing tests

**Comment by sherrif10** [comment]
cc @ibacher  @corneliouzbett  , any suggestion about this PR

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java [hunk 4]
```diff
@@ -82,6 +85,14 @@ public MethodOutcome createCondition(@ResourceParam Condition newCondition) {
 		return FhirProviderUtils.buildCreate(conditionService.saveCondition(Condition30_40.convertCondition(newCondition)));
 	}
 	
+	@Update
+	public MethodOutcome updateCondtion(@IdParam IdType id, @ResourceParam Condition newCondition) {
+		org.hl7.fhir.r4.model.Condition condition = conditionService.update(id.getIdPart(),
+		    Condition30_40.convertCondition(newCondition));
+		
+		return FhirProviderUtils.buildUpdate(newCondition);
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:93]: Thanks alot for rescuing me, i was waiting for your comments , i thought you didnt see this PR , lets do this now

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java [hunk 5]
```diff
@@ -98,4 +108,18 @@ public IBundleProvider searchConditions(
...
+			throw new ResourceNotFoundException(
+			        "Could not find medication request resource with id " + id.getIdPart() + " to delete");
+		}
+		return FhirProviderUtils.buildDelete(condition);
+	}
+	
+	public void updateCondition(IdType setValue, org.hl7.fhir.dstu3.model.Condition convertCondition) {
+		
+	}
+
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java:124]: We dont have any update method in this condition class , any suggestion please

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:93]: Thanks alot for this @ibacher , am thinking we need to implement this with convertCondition, any idea or suggestion about this thanks again for coming up

[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:93]: Finished

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java]: @ibacher  could you also take a look into this

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:93]: @sherrif10 Sorry it took me a while to get to this... I was moving through the queue. Have you made this change locally and if so, could you push it to this branch?

[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:93]: I actually changed this method and i pushed back new changes, except if you are considering it

**Comment by ibacher** [comment]
@sherrif10 The other thing I'd like to see is some tests for the `@Delete` functions.

**Comment by sherrif10** [comment]
@ibacher  sure thats what am working on

**Comment by sherrif10** [comment]
Hello sir, in this PR,  i created some methods and commented them with an intention, the mere fact is that, they seem to be collect but cause testing failure, could you point to me out the real case thank you

**Comment by sherrif10** [comment]
Hey @ibacher , Am actually sorry , there were some codes that shouldnot be here but i tried to make it more appropriate, how does it look now thanks alot @ibacher

**Comment by sherrif10** [comment]
Hello @ibacher , are we having more testings in this PR

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java [hunk 6]
```diff
@@ -79,6 +82,13 @@ public MethodOutcome createCondition(@ResourceParam Condition newCondition) {
 		return FhirProviderUtils.buildCreate(conditionService.saveCondition(newCondition));
 	}
 	
+	@Update
+	public MethodOutcome updateCondition(@IdParam IdType id, @ResourceParam Condition newCondition) {
+		org.hl7.fhir.r4.model.Condition condition = conditionService.update(id.getIdPart(), newCondition);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProvider.java:87]:
```suggestion
		Condition condition = conditionService.update(id.getIdPart(), newCondition);
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 7]
```diff
@@ -132,15 +145,35 @@ public void getConditionHistoryByWithWrongId_shouldThrowResourceNotFoundExceptio
...
-	//		assertThat(result.getCreated(), is(true));
-	//		assertThat(result.getResource(), equalTo(condition));
-	//	}
+	@Test
+	public void createCondition_shouldCreateNewCondition() {
+		when(conditionService.create(any(org.hl7.fhir.r4.model.Condition.class))).thenReturn(condition);
+		
+		MethodOutcome result = resourceProvider.createCondition(Condition30_40.convertCondition(condition));
+		
+		assertThat(result, notNullValue());
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:154]:
```suggestion
		assertThat(result, notNullValue());
		assertThat(result.getCreated(), is(true));
		assertThat(result.getResource(), equalTo(condition));
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 8]
```diff
@@ -225,6 +233,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
 		assertThat(response, isCreated());
 		assertThat(response.getStatus(), is(201));
 	}
+	//
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:236]: Can we implement these tests or at least restore them?

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_update.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "display": "John Taylor(OpenMRS ID:10010W)"
+  },
+  "recordedDate": "2019-06-18T06:37:26+03:00",
+  "recorder": {
+    "reference": "Practitioner/cf635fe4-35e6-4b9b-b3de-dc24da584150",
+    "type": "Practitioner",
+    "display": "JJ Dick"
+  }
+}
```
Comment (ibacher) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_update.json:42]:
```suggestion
}

```

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_updateWithWrongId.json [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      "value": "10010W"
+    },
+    "display": "John Taylor(OpenMRS ID:10010W)"
+  },
+  "recordedDate": "2019-06-18T06:37:26+03:00",
+  "recorder": {
+    "reference": "Practitioner/cf635fe4-35e6-4b9b-b3de-dc24da584150",
+    "type": "Practitioner",
+    "display": "JJ Dick"
+  }
```
Comment (ibacher) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_updateWithWrongId.json:40]:
```suggestion
  }
}

```

**Comment by ibacher** [comment]
@sherrif10 Some of the suggested changes don't seem to be implemented yet, and I've added a couple more comments. Could you PTAL? Thanks!

**Comment by sherrif10** [comment]
Ok thanks let me do t  that now

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_updateWithWrongId.json (hunk 10 above)
Comment (sherrif10) [reply omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_updateWithWrongId.json]: Finished, though i may encounter travis failure

**Comment by sherrif10** [comment]
Am trying to resolve these tests though you can help with  pointer, these tests have been failing for some good time, am not sure of what am missing amidst

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 11]
```diff
@@ -213,4 +245,19 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
 		assertThat(resultList.size(), greaterThanOrEqualTo(1));
 		assertThat(resultList.iterator().next().fhirType(), equalTo(FhirConstants.CONDITION));
 	}
+	
+	@Test
+	public void deleteCondtion_shouldDeletecondition() {
+
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:251]:
```suggestion
```

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_create.json [hunk 12]
```diff
@@ -15,6 +15,7 @@
       {
         "code": "116128AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
       },
+
```
Comment (varung-31) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_create.json:18]:
```suggestion
```

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_UpdateWithoutId.json [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    },
+    "display": "John Taylor(OpenMRS ID:10010W)"
+  },
+  "recordedDate": "2019-06-18T06:37:26+03:00",
+  "recorder": {
+    "reference": "Practitioner/cf635fe4-35e6-4b9b-b3de-dc24da584150",
+    "type": "Practitioner",
+    "display": "JJ Dick"
+  }
+}
```
Comment (varung-31) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ConditionResourceWebTest_UpdateWithoutId.json:41]: Please add a new line

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 14]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
...
 	
+	@Test
+	public void updateCondition_shouldUpdateCondition() {
+		when(conditionService.update(CONDITION_UUID, condition)).thenReturn(condition);
+		
+		MethodOutcome result = resourceProvider.updateCondition(new IdType().setValue(CONDITION_UUID), condition);
+		assertThat(result, notNullValue());
+		assertThat(result.getResource(), notNullValue());
+		assertThat(result.getResource().getIdElement().getIdPart(), equalTo(condition.getId()));
+
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:157]:
```suggestion
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 15]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
...
+		wrongCondition.setId(WRONG_CONDITION_UUID);
+		
+		when(conditionService.update(WRONG_CONDITION_UUID, wrongCondition)).thenThrow(MethodNotAllowedException.class);
+		resourceProvider.updateCondition(new IdType().setValue(WRONG_CONDITION_UUID), wrongCondition);
+	}
+	
+	@Test(expected = InvalidRequestException.class)
+	public void updateCondition_shouldThrowInvalidRequestExceptionForWrongConditionUuid() {
+		Condition wrongCondition = new Condition();
+		wrongCondition.setId(WRONG_CONDITION_UUID);
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:172]:
```suggestion
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 16]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
...
+		when(conditionService.update(WRONG_CONDITION_UUID, wrongCondition)).thenThrow(InvalidRequestException.class);
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:173]:
```suggestion
		when(conditionService.update(WRONG_CONDITION_UUID, condition)).thenThrow(InvalidRequestException.class);
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 17]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
...
+		
+		resourceProvider.updateCondition(new IdType().setValue(WRONG_CONDITION_UUID), wrongCondition);
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:175]:
```suggestion
		resourceProvider.updateCondition(new IdType().setValue(WRONG_CONDITION_UUID), condition);
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 18]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
...
+	public void updateCondition_shouldThrowMethodNotAllowedIfDoesNotExist() {
+		Condition wrongCondition = new Condition();
+		wrongCondition.setId(WRONG_CONDITION_UUID);
+		
+		when(conditionService.update(WRONG_CONDITION_UUID, wrongCondition)).thenThrow(MethodNotAllowedException.class);
+		resourceProvider.updateCondition(new IdType().setValue(WRONG_CONDITION_UUID), wrongCondition);
+	}
+	
+	@Test(expected = InvalidRequestException.class)
+	public void updateCondition_shouldThrowInvalidRequestExceptionForWrongConditionUuid() {
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:170]:
```suggestion
	public void updateCondition_shouldThrowInvalidRequestExceptionForUuidMismatch() {
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java [hunk 19]
```diff
@@ -143,6 +146,35 @@ public void shouldCreateNewCondition() {
 		assertThat(result.getResource(), equalTo(condition));
 	}
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:147]: Please add a test for missing id-
`updateCondition_shouldThrowInvalidRequestExceptionForMissingId`

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 20]
```diff
@@ -132,15 +145,35 @@ public void getConditionHistoryByWithWrongId_shouldThrowResourceNotFoundExceptio
...
+	}
+
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:156]: Please add a tests for missing id and not existent condition-
1. updateCondition_shouldThrowInvalidRequestExceptionForMissingId
2. updateCondition_shouldThrowMethodNotAllowedIfDoesNotExist

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 21]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+		String conditionJson;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_CONDITION_NO_ID_PATH)) {
+			Objects.requireNonNull(is);
+			conditionJson = IOUtils.toString(is, StandardCharsets.UTF_8);
+		}
+		MockHttpServletResponse response = put("/Condition/" + CONDITION_UUID).jsonContent(conditionJson)
+		        .accept(FhirMediaTypes.JSON).go();
+		assertThat(response, isBadRequest());
+		//		assertThat(response.getContentAsString(),
+		//		    containsStringIgnoringCase("Condition body must contain an ID element for update"));
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:265]: Why are these commented ?

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 22]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+	}
+	
+	@Test
+	public void updateCondition_shouldThrowExceptionForIdMissMatch() throws Exception {
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:269]:
```suggestion
	public void updateCondition_shouldThrowExceptionForIdMisMatch() throws Exception {
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 23]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_CONDITION_WRONG_ID_PATH)) {
+			Objects.requireNonNull(is);
+			conditionJson = IOUtils.toString(is, StandardCharsets.UTF_8);
+		}
+		MockHttpServletResponse response = put("/Condition/" + WRONG_CONDITION_UUID).jsonContent(conditionJson)
+		        .accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isBadRequest());
+		//		assertThat(response.getContentAsString(),
+		//		    containsStringIgnoringCase("Condition body must contain an ID element which matches the request URL"));
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:280]: Why are these commented ?

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 24]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+	}
+	
+	private boolean containsStringIgnoringCase(String string) {
+		return false;
+	}
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:285]: You don't need to use this. Instead use `Matchers.containsStringIgnoringCase()` in the above tests
```suggestion
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 25]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+		        .accept(FhirMediaTypes.JSON).go();
+		assertThat(response, isBadRequest());
+		//		assertThat(response.getContentAsString(),
+		//		    containsStringIgnoringCase("Condition body must contain an ID element for update"));
+	}
+	
+	@Test
+	public void updateCondition_shouldThrowExceptionForIdMissMatch() throws Exception {
+		String conditionJson;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_CONDITION_WRONG_ID_PATH)) {
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:271]:
```suggestion
		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_CONDITION_PATH)) {
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 26]
```diff
@@ -226,6 +234,56 @@ public void shouldCreateNewConditionGivenValidConditionResource() throws Excepti
...
+		Condition condition = new Condition();
+		condition.setIdBase(CONDITION_UUID);
+		when(conditionService.update(conditionJson, any(Condition.class))).thenReturn(condition);
+		
+		MockHttpServletResponse response = put("/Condition/" + CONDITION_UUID).jsonContent(conditionJson)
+		        .accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isOk());
+	}
+
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:253]: Please add a unit test for nonExistentCondition-
`updateCondition_shouldThrowErrorForNonExistentCondition`

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 27]
```diff
@@ -405,4 +463,31 @@ private void verifyURI(String uri) throws Exception {
 		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
 		assertThat(readBundleResponse(response).getEntry().size(), greaterThanOrEqualTo(1));
 	}
+	
+	@Test
+	public void deleteCondition_shouldDeleteCondition() throws Exception {
+		OperationOutcome outcome = new OperationOutcome();
+		outcome.setId(CONDITION_UUID);
+		outcome.getText().setDivAsString("Deleted successfully");
+
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:472]:
```suggestion
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java [hunk 28]
```diff
@@ -405,4 +463,31 @@ private void verifyURI(String uri) throws Exception {
...
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
+	}
+	
+	@Test
+	public void deleteCondition_shouldReturn404ForNonExistingCondition() throws Exception {
+		when(conditionService.get(WRONG_CONDITION_UUID)).thenReturn(null);
+		
+		MockHttpServletResponse response = get("/Condition/" + WRONG_CONDITION_UUID).accept(FhirMediaTypes.JSON).go();
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java:488]:
```suggestion
		MockHttpServletResponse response = delete("/Condition/" + WRONG_CONDITION_UUID).accept(FhirMediaTypes.JSON).go();
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 29]
```diff
@@ -212,4 +245,21 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
...
+	
+	@Delete
+	public OperationOutcome deleteObservationResource(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Condition condition = conditionService.delete(id.getIdPart());
+		if (condition == null) {
+			throw new ResourceNotFoundException("Could not find observation to delete with id" + id.getIdPart());
+		}
+		
+		return FhirProviderUtils.buildDelete(condition);
+	}
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java:257]: Thanks @varung-31 for your review lets me do this over the weekend

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java (hunk 25 above)
Comment (sherrif10) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderWebTest.java]: Finished

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java [hunk 30]
```diff
@@ -212,4 +245,21 @@ public void searchConditions_shouldReturnConditionReturnedByServiceWhenPatientIs
...
+	
+	public class UnnecessaryStubbingException extends MockitoException {
+		
+		public UnnecessaryStubbingException(String message) {
+			super(message);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirR3ResourceProviderTest.java]: Thanks @varung , finished this

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java (hunk 19 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r4/ConditionFhirResourceProviderTest.java:147]: finished thanks

**Pull request closed without merging**